eventtracking/django/models.py

```diff
@@ -0,0 +1,174 @@
+"""
+Models for filtering of events
+"""
+import logging
+import re
+
+from django.core.exceptions import ValidationError
+from django.db import models
+from model_utils.models import TimeStampedModel
+
+
+logger = logging.getLogger(__name__)
+
+
+def validate_regex_list(value):
+    """
+    Validate every regular expression in the value by trying to compile
+    it.
+
+    Raise Validation error if any reg exp is failed to compile.
+    """
+    valid_expressions, invalid_expressions = _clean_expressions(value)
+
+    if invalid_expressions:
+        error_message = 'The following invalid expressions must be corrected: {}'.format(
+            ', '.join(invalid_expressions)
+        )
+        raise ValidationError(error_message)
+
+
+def _clean_expressions(expressions_string):
+    """
+    Cleans the expressions list string by splitting using \n and then stripping
+    the whitespace characters.
+    """
+    expressions = expressions_string.split('\n')
+    raw_expressions = [expression.strip() for expression in expressions]
+    return _validate_expressions(raw_expressions)
+
+
+def _validate_expressions(expressions_list):
+    """
+    Validates every reg ex in the list and return a tuple containing
+    lists of valid and invalid expressions
+    """
+    invalid_regex_expressions = []
+    valid_regex_expressions = []
+    for exp in expressions_list:
+        try:
+            re.compile(exp)
```

Comment:
Tracing through this code I think that compile here is being called each time an event is fired then the list of filters is iterated to find a match. Because these are compiled each time this code would not gain the benefits of the compile caching. Given that the velocity in which this code would be executed using the `is_string_a_match` method I would suggest refactoring this code. I would suggest improved caching of these regexes. 

Also it seems like these may be compiled twice for each iteration of the filter? Once here to validate them and then once to compile them for checking. Code path looks like is_string_a_match -> _clean_expressions (compiled again on line 128) -> _validate_expressions (compiled first here on line 50 to validate).